File: day_12/p1.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>
#include <unordered_set>

strings Parse(const std::string& fname)
{
    return pypp::splitFileLines(fname);
}

int CountBorders(const strings& grid, int x, int y)
{
    auto count { 0 };

    for (const auto& [dx, dy] : grid::move()) {
        const auto nx = x + dx;
        const auto ny = y + dy;

        if (grid::inBounds(grid, nx, ny)) {
            if (grid[nx][ny] != grid[x][y])
                count++;
        } else
            count++;
    }

    return count;
}

std::unordered_map<char, std::vector<std::vector<grid::Location>>> RegionsOf(const strings& grid)
{
    const auto n = grid.size();
    const auto m = grid[0].length();

    std::unordered_map<char, std::vector<std::vector<grid::Location>>> regions_map {};
    std::unordered_set<grid::Location, collections::TupleHash> visited {};
    for (std::size_t i { 0U }; i < n; ++i) {
        for (std::size_t j { 0U }; j < m; ++j) {
            if (visited.find({ i, j }) == visited.end()) {
                const auto label = grid[i][j];
                const grid::Location start = { i, j };
                auto condition = [&label](const char ch) { return ch == label; };
                const auto locations = algorithms::dfs(grid, start, condition, visited);
                regions_map[label];
                if (!locations.empty())
                    regions_map[label].push_back(locations);
            }
        }
    }

    return regions_map;
}

long Price(const strings& grid, const std::unordered_map<char, std::vector<std::vector<grid::Location>>>& regions_map)
{
    auto price { 0 };

    for (const auto [label, locations_list] : regions_map) {
        for (const auto locations : locations_list) {
            const auto area = locations.size();
            auto perimeter { 0 };
            for (const auto [x, y] : locations)
                perimeter += CountBorders(grid, x, y);
            price += area * perimeter;
        }
    }

    return price;
}

long Compute(const strings& grid)
{
    const auto regions_map = RegionsOf(grid);
    return Price(grid, regions_map);
}

int main()
{
    const auto grid = Parse("input.txt");
    std::cout << Compute(grid) << std::endl;

    return 0;
}

File: day_12/p1.py
from typing import Dict, Generator, List, Tuple

def parse(fname: str) -> List[str]:
    with open(fname, 'r') as f:
        return f.read().splitlines()

def move_gen(x: int, y: int) -> Generator[Tuple[int, int], None, None]:
    yield (x - 1, y)  # up
    yield (x + 1, y)  # down
    yield (x, y + 1)  # left
    yield (x, y - 1)  # right

def is_in_bounds(grid: List[str], x: int, y: int) -> bool:
    return x in range(len(grid)) and y in range(len(grid[0]))

def how_many_borders(i: int, j: int, grid: List[str]) -> int:
    count = 0
    for ni, nj in move_gen(i, j):
        if is_in_bounds(grid, ni, nj):
            if grid[ni][nj] != grid[i][j]:
                count += 1
        else:
            count += 1

    return count

def regions_from(grid: List[str]) -> Dict[str, List[List[Tuple[int, int]]]]:
    n, m = len(grid), len(grid[0])
    seen = set()

    def dfs(label, i, j):
        lot = []
        stack = [(i, j)]

        while stack:
            cx, cy = stack.pop()
            if (cx, cy) not in seen and grid[cx][cy] == label:
                seen.add((cx, cy))
                lot.append((cx, cy))

                for nx, ny in move_gen(cx, cy):
                    if is_in_bounds(grid, nx, ny) and (nx, ny) not in seen:
                        stack.append((nx, ny))

        return lot

    regions: Dict[str, List[List[Tuple[int, int]]]] = {}
    for i in range(n):
        for j in range(m):
            if (i, j) not in seen:
                current_label = grid[i][j]
                lot = dfs(current_label, i, j)
                if lot:
                    regions.setdefault(current_label, []).append(lot)

    return regions

def prices_of(regions: Dict[str, List[List[Tuple[int, int]]]], grid: List[str]) -> int:
    price = 0
    for _, lots_list in regions.items():
        for lots in lots_list:
            area = len(lots)
            perimeter = 0
            for lot in lots:
                perimeter += how_many_borders(lot[0], lot[1], grid)
            price += area * perimeter

    return price

def compute(grid: List[str]) -> int:
    regions = regions_from(grid)
    return prices_of(regions, grid)

def main() -> None:
    grid = parse("input.txt")
    print(compute(grid))


if __name__ == "__main__":
    main()
